File: docs/courses/data-structure/samples/stack/LinkedListBaseStack/LinkedListBaseStack.py
class Node:
    def __init__(self, value=None, link=None):
        self.value = value
        self.next = link

    def __str__(self):
        return str(self.value)
        


class LinkedListBaseStack:
    def __init__(self):
        self.head = None

    def __iter__(self):
        v = self.head
        while v is not None:
            yield v
            assert isinstance(v.next, object)
            v = v.next
    
    def __str__(self):
        return " -> ".join(str(v) for v in self)
    
    def isEmpty(self):
        if self.head:
            return False
        return True
    
    def push(self, data):
        node = Node(data)
        node.next = self.head
        self.head = node

    def pop(self):
        if self.isEmpty():
            return -1
        data = self.head.value
        self.head = self.head.next

        return data
    
    def peek(self):
        if self.isEmpty():
            return -1

        return self.head.value

# 테스트
stack = LinkedListBaseStack()

stack.push(522)
stack.push("BoanLab")
stack.push("Apdul")
stack.push("LinkedListBaseStack")

'''
expected output
LinkedListBaseStack
Apdul
BoanLab
522
'''
while not stack.isEmpty():
    print(stack.pop())

File: docs/courses/data-structure/samples/stack/LinkedListBaseStack/LinkedListBaseStack.java
package LinkedListBaseStack;

import java.util.NoSuchElementException;

public class LinkedListBaseStack {
    private static class Node {
        private Object data;
        private Node next;

        public Node(Object data) {
            this.data = data;
            this.next = null;
        }
    }
    private Node top;

    public LinkedListBaseStack() {
        this.top = null;
    }

    public boolean isEmpty() {
        return top == null;
    }

    public void push(Object data) {
        Node node = new Node(data);
        node.next = top;
        top = node;
    }

    public Object peek() {
        return top.data;
    }

    public Object pop() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        } else {
            Object data = top.data;
            top = top.next;

            return data;
        }
    }
}